community/collections/src/main/java/org/neo4j/helpers/collection/Iterators.java

```diff
@@ -724,4 +729,44 @@ public static String join( String joinString, Iterator<?> iter )
         }
         return sb.toString();
     }
+
+    /**
+     * Create a stream from the given iterator.
+     * <p>
+     * <b>Note:</b> returned stream needs to be closed via {@link Stream#close()} if the given iterator implements
+     * {@link Resource}.
+     *
+     * @param iterator the iterator to convert to stream
+     * @param <T> the type of elements in the given iterator
+     * @return stream over the iterator elements
+     * @throws NullPointerException when the given stream is {@code null}
+     */
+    public static <T> Stream<T> stream( Iterator<T> iterator )
+    {
+        return stream( iterator, 0 );
+    }
+
+    /**
+     * Create a stream from the given iterator with given characteristics.
+     * <p>
+     * <b>Note:</b> returned stream needs to be closed via {@link Stream#close()} if the given iterator implements
+     * {@link Resource}.
+     *
+     * @param iterator the iterator to convert to stream
+     * @param characteristics the logical OR of characteristics for the underlying {@link Spliterator}
+     * @param <T> the type of elements in the given iterator
+     * @return stream over the iterator elements
+     * @throws NullPointerException when the given iterator is {@code null}
+     */
+    public static <T> Stream<T> stream( Iterator<T> iterator, int characteristics )
+    {
+        Objects.requireNonNull( iterator );
+        Spliterator<T> spliterator = Spliterators.spliteratorUnknownSize( iterator, characteristics );
+        Stream<T> stream = StreamSupport.stream( spliterator, false );
```

Comment:
I do not think we can say anything about spliterator characteristics in such generic method. It takes `Iterable` and we do not know if it is `IMMUTABLE` or `CONCURRENT`.

Late binding influences ability of the spliterator to spot changes of the source before first traversal, split or size estimate. In our case, I guess, users are not supposed to modify the given iterable after construction of the stream.

Maybe we should start with this simple implementation and enhance it if performance problems occur?